src/app/codegen-data-model/CodegenDataModel_Read.cpp

```diff
@@ -0,0 +1,350 @@
+    case ZCL_INT32S_ATTRIBUTE_TYPE: // Signed 32-bit integer
+        return EncodeFromSpan<int32_t>(data, isNullable, encoder);
+    case ZCL_INT40S_ATTRIBUTE_TYPE: // Signed 40-bit integer
+        return EncodeFromSpan<OddSizedInteger<5, true>>(data, isNullable, encoder);
+    case ZCL_INT48S_ATTRIBUTE_TYPE: // Signed 48-bit integer
+        return EncodeFromSpan<OddSizedInteger<6, true>>(data, isNullable, encoder);
+    case ZCL_INT56S_ATTRIBUTE_TYPE: // Signed 56-bit integer
+        return EncodeFromSpan<OddSizedInteger<7, true>>(data, isNullable, encoder);
+    case ZCL_INT64S_ATTRIBUTE_TYPE: // Signed 64-bit integer
+        return EncodeFromSpan<int64_t>(data, isNullable, encoder);
+    case ZCL_SINGLE_ATTRIBUTE_TYPE: // 32-bit float
+        return EncodeFromSpan<float>(data, isNullable, encoder);
+    case ZCL_DOUBLE_ATTRIBUTE_TYPE: // 64-bit float
+        return EncodeFromSpan<double>(data, isNullable, encoder);
+    case ZCL_CHAR_STRING_ATTRIBUTE_TYPE: // Char string
+        return EncodeStringLike<CharSpan, ShortPascalString>(data, isNullable, encoder);
+    case ZCL_LONG_CHAR_STRING_ATTRIBUTE_TYPE:
+        return EncodeStringLike<CharSpan, LongPascalString>(data, isNullable, encoder);
+    case ZCL_OCTET_STRING_ATTRIBUTE_TYPE: // Octet string
+        return EncodeStringLike<ByteSpan, ShortPascalString>(data, isNullable, encoder);
+    case ZCL_LONG_OCTET_STRING_ATTRIBUTE_TYPE:
+        return EncodeStringLike<ByteSpan, LongPascalString>(data, isNullable, encoder);
+    default:
+        ChipLogError(DataManagement, "Attribute type 0x%x not handled", static_cast<int>(metadata->attributeType));
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedRead);
+    }
+}
+
+} // namespace
+
+/// separated-out ReadAttribute implementation (given existing complexity)
+///
+/// Generally will:
+///    - validate ACL (only for non-internal requests)
+///    - Try to read attribute via the AttributeAccessInterface
+///    - Try to read the value from ember RAM storage
+CHIP_ERROR CodegenDataModel::ReadAttribute(const InteractionModel::ReadAttributeRequest & request, AttributeValueEncoder & encoder)
+{
+    ChipLogDetail(DataManagement,
+                  "Reading attribute: Cluster=" ChipLogFormatMEI " Endpoint=%x AttributeId=" ChipLogFormatMEI " (expanded=%d)",
+                  ChipLogValueMEI(request.path.mClusterId), request.path.mEndpointId, ChipLogValueMEI(request.path.mAttributeId),
+                  request.path.mExpanded);
+
+    // ACL check for non-internal requests
+    if (!request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
+    {
+        ReturnErrorCodeIf(!request.subjectDescriptor.has_value(), CHIP_ERROR_INVALID_ARGUMENT);
+
+        Access::RequestPath requestPath{ .cluster = request.path.mClusterId, .endpoint = request.path.mEndpointId };
+        ReturnErrorOnFailure(Access::GetAccessControl().Check(*request.subjectDescriptor, requestPath,
```

Comment:
Spec requires _very_ particular behavior here depending on whether path was expanded or not.  That needs to be handled here, because caller will not be able to tell reliably that this was a failed ACL check as opposed to some other failure.  And no, sniffing the exact value of the CHIP_ERROR is not reliable.

Alternately, we need a way to communicate to the caller that this failed because of the ACL check....

Reply:
Would it be acceptable to have a dedicated error for these checks?

using path.mExpanded is strange and I would claim wrong: it is not obvious that mExpanded means `ACL check is not done` unless one reads comments carefully and if you need to read comments carefully, this means it is significantly harder to maintain.

The user  that actually decides if ACL failures or unsupported read is acceptable seems to be the caller (subscription with wildcard or not), so we should either:
   - get this call to Read tell us if this failed ACL (and caller decides how to handle the error) - this is the path I took
   - pass in to the caller if you want read be ok with ACL failure via some argument. I have some complains about how ember-compatibility-functions does this today:
       - mExpanded is not a reasonably named flag for this "ignore acl checks" logic
       - returning CHIP_NO_ERROR but not actually filling the read value is odd. I would expect "if you do not get a read error, the read buffer is filled with appropriate data". Given that null and empty values can be valid, just returning CHIP_NO_ERROR does not seem reasonable.

Is there an approach that you would suggest? I would like one that addresses my concerns with returning CHIP_NO_ERROR.

Reply:
> Would it be acceptable to have a dedicated error for these checks?

You mean a dedicated CHIP_ERROR value?  I would really rather not.  People use CHIP_ERROR values willy-nilly based on what they sound like.

If we are trying to communicate a specific state, we should communicate that state.  If we need a richer return type than CHIP_ERROR, we should do that.

> using path.mExpanded is strange and I would claim wrong

Just to be clear, the only reason mExpanded exists is to allow this check to be done, yes?  It affects more than just ACL checks, by the way, which is why it's named like it is.

> unless one reads comments carefully

Or actually reads the spec.

> The user that actually decides if ACL failures or unsupported read is acceptable seems to be the caller
> (subscription with wildcard or not)

It's not just subscriptions.  It's reads too.

I think a fundamental issue here seems to be that you are reading the ember-compatibility-functions code, trying to guess what it's trying to do, and then trying to reimplement that, instead of actually implementing the spec.  Reverse-engineering the spec from the code is much harder than reading the spec, and hence more error-prone.

> get this call to Read tell us if this failed ACL (and caller decides how to handle the error) - this is the path I took

Again, it's not just ACL checks.

> mExpanded is not a reasonably named flag for this "ignore acl checks" logic

That's because it affects things other than ACL checks.  The relevant spec bits say:

> if the path is a concrete path:
> (acl checks, existence checks, and the like here)
>Else perform Request Path Expansion and process each expanded existent path as follows:
>
>   If the path indicates attribute data that is not readable, then the path SHALL be discarded.
>
>    Else if reading from the attribute in the path requires a privilege that is not granted to access the cluster in the path, then the path SHALL be discarded.
>
>each remaining error free existent path is processed as follows: (actually do reads here)

That's the actual logic that needs to be implemented.  The way it's handled in ember-compatibility-functions is a specific way to implement it that it pushes the "this was not a concrete path" boolean down into the the read and does all the checks inside the read.  This means it re-does existence checks for things we got from wildcard expansion, which is a bit silly, but keeps the codepaths down.  But we could adopt whatever implementation strategy we want here.

Reply:
Updated code to keep logic from ember-compatibility-functions and added unit tests for this.